update ResourceMetricsAPI tests

MHBauer opened the pull request:
/metrics/resource/v1alpha1 was deprecated and moved to
/metrics/resource

Renames to remove v1alpha1 from function names and matcher variables.

Pod deletion was taking multiple minutes, so set GracePeriodSeconds to 0.

Removed restart loop during test pod startup.

Move ResourceMetricsAPI out of Orphans by giving it a NodeFeature tag.

API removed in 7b7c73bf8754f2d9c7c3bfa7f3a196fdb19e74f1 #88568
Test created 6051664c0fb0eb78a84cea8bf5a1e1e4584de50e #73946

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind cleanup
/kind failing-test

**What this PR does / why we need it**:
API moved without updating test.

**Which issue(s) this PR fixes**:
Failing tests in https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-orphans

**Special notes for your reviewer**:
/cc @vpickard 
/assign @dashpole 
Is there an explanation for the restart loop? It looks like it may have been a copy and paste from the garbage collection test.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

hasheddan commented on test/e2e_node/resource_metrics_test.go:
```diff
@@ -41,36 +40,24 @@ const (
 			time.Sleep(15 * time.Second)
 		})
-		ginkgo.It("should report resource usage through the v1alpha1 resouce metrics api", func() {
+		ginkgo.It("should report resource usage through the resouce metrics api", func() {
```
nitpick:
```suggestion
		ginkgo.It("should report resource usage through the resource metrics api", func() {
```

MHBauer replied:
thank you.

SergeyKanzhelev commented:
Just for my understanding, I see that the alpha endpoint is still being registered:

https://github.com/kubernetes/kubernetes/blob/da5ec168e71d199721bab32caee565386ff47306/pkg/kubelet/server/server.go#L381-L387

Were tests failing because of a timeout?

As for the change, even if alpha endpoint is still registered, I don't think it's worth validating it and the change looks good:

/lgtm
/retest

SergeyKanzhelev commented:
looks like bazel update is needed.

dashpole commented on test/e2e_node/resource_metrics_test.go:
```diff
@@ -41,36 +40,24 @@ const (
-			ginkgo.By("Waiting for test pods to restart the desired number of times")
-			gomega.Eventually(func() error {
-				for _, pod := range pods {
-					err := verifyPodRestartCount(f, pod.Name, len(pod.Spec.Containers), numRestarts)
```
This was initially added to catch a bug in which container metrics would disappear after container restarts.  It was copied from the summary API test IIRC.  I'd like to keep this if possible.

MHBauer replied:
I put back the loop with your explanation as comment.
I'm not sure this is going to have the desired effect since we wait later for the metrics to populate. WDYT?

dashpole replied:
Right.  IIRC, the bug that we fixed, and added this test for was where we only collected metrics for the first instance of a container, and not for ones created after the container restarts.  So we want to first have a pod with containers that have already restarted, and then make sure metrics look correct after that.  Make sense?

MHBauer replied:
I understand now. I'll think on this and if it needs to be rewritten again. Thank you.

MHBauer replied:
I think this should be 
BY "restarting containers to ensure that future instances of a container besides the first instance collect metrics"
or something like that. 

Is there an issue/PR we can link to for demonstative purposes? Or does this need a separate Test case itself?

MHBauer commented on test/e2e_node/resource_metrics_test.go:
```diff
@@ -64,13 +63,13 @@ var _ = framework.KubeDescribe("ResourceMetricsAPI", func() {
 		})
-		ginkgo.It("should report resource usage through the v1alpha1 resouce metrics api", func() {
+		ginkgo.It("should report resource usage through the resource metrics api", func() {
 			ginkgo.By("Fetching node so we can know proper node memory bounds for unconstrained cgroups")
 			node := getLocalNode(f)
 			memoryCapacity := node.Status.Capacity["memory"]
 			memoryLimit := memoryCapacity.Value()
```
@dashpole can you help me understand this comment? It looks like it refers to the following couple of lines, and then goes into the matcher. I'm guessing that this is sort of a default return value of the node having access to all of it's own memory?

dashpole replied:
The matcher checks that the returned value is within a range.  Memory usage should not exceed the capacity of the node, so it is an upper bound on some ranges IIRC.

MHBauer reviewed (commented):
rebased + hack/update-bazel.sh

> Just for my understanding, I see that the alpha endpoint is still being registered:
> 
> https://github.com/kubernetes/kubernetes/blob/da5ec168e71d199721bab32caee565386ff47306/pkg/kubelet/server/server.go#L381-L387
> 
> Were tests failing because of a timeout?
> 
> As for the change, even if alpha endpoint is still registered, I don't think it's worth validating it and the change looks good:

Tests were failing because there was no content in the response. I think #88568 shows that the endpoint still exists, but has nothing in it, do I also need to remove the registration in `pkg/kubelet/server/server.go` ? 

```
_output/local/go/src/k8s.io/kubernetes/test/e2e_node/resource_metrics_test.go:67
Timed out after 60.000s.
Expected
    <string>: KubeletMetrics
to match keys: {
missing expected key scrape_error
missing expected key node_cpu_usage_seconds_total
missing expected key node_memory_working_set_bytes
missing expected key container_cpu_usage_seconds_total
missing expected key container_memory_working_set_bytes
}
```

Maybe I've gone about this the wrong way and we should put back the removed api? Or remove even more code?

dashpole commented:
Oh!  I didn't realize the resource metrics test was failing

We should definitely fix it in 1.19.  The existing endpoint should still work this release.  cc @RainbowMango

dashpole commented:
Would it be possible to test both endpoints?  I'd like to try and fix the v1alpha1 endpoint

MHBauer commented on test/e2e_node/resource_metrics_test.go:
```diff
@@ -64,13 +63,13 @@ var _ = framework.KubeDescribe("ResourceMetricsAPI", func() {
 			ginkgo.By("Waiting 15 seconds for cAdvisor to collect 2 stats points")
 			time.Sleep(15 * time.Second)
 		})
-		ginkgo.It("should report resource usage through the v1alpha1 resouce metrics api", func() {
```
Add another clause like this for v1alpha1? Same test, different endpoints?

dashpole replied:
Yes.  #88568 was supposed to keep the contents of the endpoint the same IIRC, but just refactor and add a new endpoint.

RainbowMango replied:
> Yes. #88568 was supposed to keep the contents of the endpoint the same IIRC, but just refactor and add a new endpoint.

Yes.

MHBauer reviewed (commented):
@dashpole Does that mean add back the code of https://github.com/kubernetes/kubernetes/pull/88568 ? 

 Same test, different endpoints?

MHBauer commented on pkg/kubelet/server/server.go:
```diff
@@ -378,20 +378,17 @@ func (s *Server) InstallDefaultHandlers(enableCAdvisorJSONEndpoints bool) {
+	s.addMetricsBucketMatcher("metrics/resource/v1alpha1")
+	s.restfulCont.Handle(path.Join(resourceMetricsPath, v1alpha1.Version),
+		compbasemetrics.HandlerFor(resourceRegistry, compbasemetrics.HandlerOpts{ErrorHandling: compbasemetrics.ContinueOnError}),
+	)
```
I think this is what we want to do? Hook up the alpha endpoint to the stable endpoint.

Alpha endpoint had to move below the existing one to delete the registry and pick up the stable registry endpoint.

RainbowMango replied:
> I think this is what we want to do?

No, what we want to do is DO NOT touch the original endpoint `metrics/resource/v1alpha1` and add a new endpoint to replace the original one.

MHBauer commented on test/e2e_node/resource_metrics_test.go:
```diff
@@ -89,15 +89,19 @@ var _ = framework.KubeDescribe("ResourceMetricsAPI", func() {
 			})
+			ginkgo.By("ensuring both resource metrics APIs return the same values")
 			ginkgo.By("Giving pods a minute to start up and produce metrics")
-			gomega.Eventually(getV1alpha1ResourceMetrics, 1*time.Minute, 15*time.Second).Should(matchV1alpha1Expectations)
+			gomega.Eventually(getResourceMetrics, 1*time.Minute, 15*time.Second).Should(matchResourceMetrics)
+			gomega.Eventually(getV1alpha1ResourceMetrics, 1*time.Minute, 15*time.Second).Should(matchResourceMetrics)
 			ginkgo.By("Ensuring the metrics match the expectations a few more times")
-			gomega.Consistently(getV1alpha1ResourceMetrics, 1*time.Minute, 15*time.Second).Should(matchV1alpha1Expectations)
+			gomega.Consistently(getResourceMetrics, 1*time.Minute, 15*time.Second).Should(matchResourceMetrics)
+			gomega.Consistently(getV1alpha1ResourceMetrics, 1*time.Minute, 15*time.Second).Should(matchResourceMetrics)
```
I don't know ginkgo well enough to know if we can do two comparisons in one call here. 

This serializes everything, so it takes 2x as long since we wait a minute for each.

dashpole replied:
I'm fine with this, or i'm also fine having separate `If(` blocks, and making the entire test a function that takes the getResourceMetrics() function.  The advantage of the latter is that it is easy to tell which endpoint isn't working.

MHBauer replied:
I'll mess around with it some more and see if I can split it up without introducing a bunch of duplication.

MHBauer replied:
I need to modify the matcher to for each case to ignore the other's pods. Not sure if there is a namespaced call to resource API that this should use instead.

MHBauer reviewed (commented):
Checking to make sure I have the right idea here.
 /cc  @RainbowMango
as the original mover of some of this code.

MHBauer commented:
That's an interesting failure, 
```
I0813 21:07:20.026] STEP: exposing timestamps
I0813 21:07:20.027] Aug 13 21:07:00.471: INFO: Running '/go/src/k8s.io/kubernetes/bazel-bin/cmd/kubectl/linux_amd64_pure_stripped/kubectl --server=https://[::1]:46649 --kubeconfig=/root/.kube/kind-test-config logs logs-generator logs-generator --namespace=kubectl-9623 --tail=1 --timestamps'
I0813 21:07:20.027] Aug 13 21:07:01.947: INFO: stderr: ""
I0813 21:07:20.027] Aug 13 21:07:01.947: INFO: stdout: "2020-08-13T21:07:01.709817451Z I0813 21:07:01.699522       1 logs_generator.go:76] 36 GET /api/v1/namespaces/kube-system/pods/csmw 221\n2020-08-13T21:07:01.897899022Z I0813 21:07:01.890402       1 logs_generator.go:76] 37 GET /api/v1/namespaces/default/pods/2t6 559\n"
I0813 21:07:20.028] Aug 13 21:07:01.947: INFO: got output "2020-08-13T21:07:01.709817451Z I0813 21:07:01.699522       1 logs_generator.go:76] 36 GET /api/v1/namespaces/kube-system/pods/csmw 221\n2020-08-13T21:07:01.897899022Z I0813 21:07:01.890402       1 logs_generator.go:76] 37 GET /api/v1/namespaces/default/pods/2t6 559\n"
I0813 21:07:20.028] Aug 13 21:07:01.948: FAIL: Expected
I0813 21:07:20.028]     <int>: 2
I0813 21:07:20.028] to equal
I0813 21:07:20.028]     <int>: 1 
```
Don't think it's related to this change, but maybe a bug in `kubectl get logs` to investigate.

MHBauer commented:
looks like flake #92468

MHBauer commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

RainbowMango commented:
> We should definitely fix it in 1.19. The existing endpoint should still work this release. 

Yes. In 1.19 both `metrics/resource/v1alpha1` and `metrics/resource` are available, but as per metrics deprecation policy, no metrics will be emitted from `metrics/resource/v1alpha1` by default, because all metrics in this endpoint will become to `hidden`.

MHBauer commented:
> > We should definitely fix it in 1.19. The existing endpoint should still work this release.
> 
> Yes. In 1.19 both `metrics/resource/v1alpha1` and `metrics/resource` are available, but as per metrics deprecation policy, no metrics will be emitted from `metrics/resource/v1alpha1` by default, because all metrics in this endpoint will become to `hidden`.

What does that mean here? Is there a way to unhide them?

RainbowMango commented on test/e2e_node/resource_metrics_test.go:
```diff
@@ -84,20 +84,30 @@ var _ = framework.KubeDescribe("ResourceMetricsAPI", func() {
+			gomega.Consistently(fetcher, 1*time.Minute, 15*time.Second).Should(matchResourceMetrics)
+		}
+
+		ginkgo.It("should report resource usage through the alpha resource metrics api", func() {
+			resourceFetcher := getV1alpha1ResourceMetrics
+			fetchAndTest(f, resourceFetcher)
+		})
+		ginkgo.It("should report resource usage through the stable resource metrics api", func() {
+			resourceFetcher := getResourceMetrics
+			fetchAndTest(f, resourceFetcher)
```
We can't say the `/metrics/resource/v1alpha1` is the alpha resource metrics  API and `/metrics/resource` is not the stable resource metrics API, this will make confusion. They are both the resource metrics API in different endpoints.

RainbowMango commented on test/e2e_node/resource_metrics_test.go:
```diff
@@ -111,9 +121,15 @@ var _ = framework.KubeDescribe("ResourceMetricsAPI", func() {
 })
 
 func getV1alpha1ResourceMetrics() (e2emetrics.KubeletMetrics, error) {
```
Just rename function name to `getResourceMetrics` will be fine.

MHBauer replied:
yea, that's an earlier version

RainbowMango reviewed (commented):
Given all metrics under `/metrics/resources/v1alpha1` now become to `hidden` state by default and will be removed soon. So I think we can migrate the E2E test to `/metrics/resources`.

RainbowMango commented:
> What this PR does / why we need it:
API moved without updating test.

@MHBauer The API didn't move, what we did is added a new endpoint(/metrics/resources) and deprecated all metrics under the old endpoint(/metrcis/resources/v1alpha1). 
I didn't realize we still have E2E test which cares about the resources metrics, sorry for that.

To sum up, what we need to do is update E2E test to validate the metrics from the new endpoint.

RainbowMango commented:
> Would it be possible to test both endpoints? I'd like to try and fix the v1alpha1 endpoint

I doubt it worth doing so. We are going to remove the v1alpha1 endpoint soon. So, I suggest stopping test the v1alpha1 endpoint.

SergeyKanzhelev commented:
> > Would it be possible to test both endpoints? I'd like to try and fix the v1alpha1 endpoint
> 
> I doubt it worth doing so. We are going to remove the v1alpha1 endpoint soon. So, I suggest stopping test the v1alpha1 endpoint.

Should it be release noted in 1.19 then that endpoint was deprecated?

RainbowMango commented:
> Should it be release noted in 1.19 then that endpoint was deprecated?

@SergeyKanzhelev 
We have noted the deprecation in v1.18 release note by #86282, I guess we don't need to note again.
```
kubelet: resource metrics endpoint `/metrics/resource/v1alpha1` as well as all metrics under this endpoint have been deprecated.
Please convert to the following metrics emitted by endpoint `/metrics/resource`:
```

dashpole commented:
Ah, thanks @RainbowMango.  I forgot metrics auto-hide a release after being deprecated...

I agree we don't need to test the v1alpha1 endpoint anymore.  We can also limit this PR to just updating the endpoint for the test.

MHBauer commented:
Okay, summarizing
1. don't enable the alpha api
2. fix the test for the regular api only.

SergeyKanzhelev commented:
Should then v1alpha1  endpoint registration be removed as well? Maybe a separate PR?

MHBauer commented:
Should I also remove the v1alpha1 api registration?

MHBauer committed:
update ResourceMetricsAPI node-e2e test

/metrics/resource/v1alpha1 was deprecated and moved to
/metrics/resource

Renames to remove v1alpha1 from function names and matcher variables.

Pod deletion was taking multiple minutes, so set GracePeriodSeconds to 0.

Commented restart loop during test pod startup.

Move ResourceMetricsAPI out of Orphans by giving it a NodeFeature tag.

API removed in 7b7c73b #88568
Test created 6051664 #73946

dashpole commented:
> Should then v1alpha1 endpoint registration be removed as well? Maybe a separate PR?

> Should I also remove the v1alpha1 api registration?

After 1.19 is cut.

dashpole commented:
We still need the endpoint, as metrics can be unhidden by flag.

MHBauer commented:
What's the flag to unhide? I am curious, more than anything

MHBauer commented:
alright, this is hopefully the final version 🤞

MHBauer commented:
I don't know what to do with this:
```
failed to try resolving symlinks in path "/var/log/pods/test-pods_09debb56-de5c-11ea-bf0d-0a7878750357_0b856363-de5c-11ea-9893-42010a800098/test/0.log": lstat /var/log/pods/test-pods_09debb56-de5c-11ea-bf0d-0a7878750357_0b856363-de5c-11ea-9893-42010a800098/test/0.log: no such file or directory 
```
/test pull-kubernetes-e2e-gce-100-performance

dashpole commented:
the flag is (e.g.) `--show-hidden-metrics-for-version=1.19` to re-enable these disabled metrics.

/lgtm
/approve

MHBauer commented:
looks like maybe a rare flake? 18 in the past week https://storage.googleapis.com/k8s-gubernator/triage/index.html?test=Services%20should%20be%20able%20to%20change%20the%20type%20from%20NodePort%20to%20ExternalName#2fa91b3a0e6e1a4514d4
/test pull-kubernetes-conformance-kind-ipv6-parallel

MHBauer commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

MHBauer commented:
alrighty, looking forward to thaw.

RainbowMango commented on test/e2e_node/resource_metrics_test.go:
```diff
@@ -110,8 +110,9 @@ var _ = framework.KubeDescribe("ResourceMetricsAPI", func() {
-func getV1alpha1ResourceMetrics() (e2emetrics.KubeletMetrics, error) {
-	return e2emetrics.GrabKubeletMetricsWithoutProxy(framework.TestContext.NodeName+":10255", "/metrics/resource/"+kubeletresourcemetricsv1alpha1.Version)
+func getResourceMetrics() (e2emetrics.KubeletMetrics, error) {
+	ginkgo.By("getting stable resource metrics API")
```
nit: remove `stable`, or refactor the log with something like `grabbing kubelet resource metrics`.

RainbowMango commented:
@MHBauer 
v1.19 has been released and #94272 will clean up the deprecated `/metrics/resource/v1alpha1` endpoint.

MHBauer commented:
Looks like the matcher needs to be tweaked and/or the test marked as disruptived as it can be failing due to the local environment.
